Force selection of calendar or fixed intervals in date histo agg

polyfractal opened the pull request:
*WIP: still needs documentation, and probably some more randomization tests*

This adds a `calendar_interval` and `fixed_interval` parameter to the date histogram aggregation, as well as deprecating the `interval` parameter.  The two new parameters are mutually exclusive, and cannot be set with `interval` either.  It also deprecates the `interval()` and `dateHistogramInterval()` methods of the builder.   Finally, it adjusts the composite agg and rollup's to use a similar scheme.   

This makes the selection of calendar vs fixed time units explicit to the user, and provides immediate validation of the unit they provide.

This PR should be fully BWC, and throws deprecation warnings where possible. 

Some notable points which may need adjusting:
- If the user opts to use the new parameters, we will accept any valid unit even if it "overlaps".  For example, `1h` can be both a fixed and calendar unit.  If the user specifies `fixed_interval: 1h`, we know it is a fixed hour, as opposed to `calendar_interval: 1h` which indicates calendar-aware hour.

    If the user continues to use the deprecated `interval: 1h`, we fall back to the old behavior where we assume units are calendar if possible.

    This felt like the "right" way to do it, rather than forcing the user to understand that a fixed hour had to be specified as `60m`, etc.

- DateHisto agg builder was a mess, since it accepted both a fixed `long` interval and a `DateHistogramInterval` which could be fixed or cal.  And now we have two more parameters.  Lastly, this logic needed to duplicated to composite agg too.  To keep track of this (and reduce copy/paste between composite agg), there's a new `DateIntervalWrapper` which records which method is used and enforces which can be set together or not.

    Internally, the wrapper gets rid of the `long` interval and records everything as a DateHistogramInterval.  Once the deprecation period is up this class will be greatly simplified. 

- The rollup date histo config just got rid of builders during HLRC refactoring.  But now we have `interval` or `one_of(fixed_interval, calendar_interval)`, which doesn't play nicely with constructing object parsers.  So I opted to specialize the DateHistoGroupConfig object into fixed/calendar objects.  They just do simple validation and provide a way to tell the configs apart at a later point.

    This let us avoid having weird ctor arguments where two of the three are null (`interval, fixed_interval, calendar_interval`), and made it easier to deprecate the old way.  Happy to entertain a different idea if it's better.

/cc @colings86 @jimczi

elasticmachine commented:
Pinging @elastic/es-search-aggs

polyfractal commented:
Note: rollup still stores the the interval as `interval` in rollup docs, although we should move to also storing `fixed_interval`/`calendar_interval`.  But since this PR is already enormous I'm leaving that for later.

polyfractal commented:
Jenkins, run gradle build tests

polyfractal commented:
@elasticmachine run elasticsearch-ci/packaging-sample

polyfractal commented:
@elasticmachine run elasticsearch-ci/packaging-sample

polyfractal commented:
@elasticmachine run elasticsearch-ci/packaging-sample

polyfractal commented:
@elasticmachine run elasticsearch-ci/2

polyfractal commented:
@elasticmachine test this please

polyfractal commented:
@elasticmachine run elasticsearch-ci/1

Commit:
Core code changes

Commit:
Rollup code and tests

Commit:
ML code and tests

Commit:
Other tests

Commit:
Docs

polyfractal commented:
Status update!  I think the majority of test failures have been squashed, so I've re-arranged the commits into rough groupings: 

- core code changes
- Rollup and tests
- ML and tests
- Other tests
- docs.  

Hopefully helps make reviewing this a little easier.

It's been some time since I've looked at the code changes itself, so I'm going to review those again and make sure everything looks good, and I'll be requisitioning some eyeballs from ML for the ML-related changes.  

Where it made sense I updated ML (e.g. if it was using `long milliseconds` I just converted to `fixed_time`), but in many locations I didn't want to break code so I left the deprecated settings in place and adjusted tests to deal with the deprecation warnings.  Some tests were impossible to fix though, so deprecations were disabled.

polyfractal commented:
@colings86 Ok, I think the core code changes OK for review.  I still need to vet the rest of the test changes for sanity, will take care of that early next week.

I'd probably start with `DateIntervalWrapper` to get a feel for how things change, since that's where the bulk of the logic takes place.  Everything else is replacing current interval parsing with `DateIntervalWrapper`

colings86 reviewed (commented):
Reviewed the first commit and left comments, will review each commit in separate passes

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java:
```diff
@@ -0,0 +1,415 @@
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            IntervalTypeEnum type = this;
```
Weird, not sure.  Doesn't look necessary, think it's a leftover artifact of a different revision.

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java:
```diff
@@ -0,0 +1,415 @@
+        if (intervalType.equals(IntervalTypeEnum.CALENDAR) || tryIntervalAsCalendarUnit() != null) {
+            return dateHistogramInterval;
+        }
+        return null;
```
++ I think we can do that.

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java:
```diff
@@ -0,0 +1,415 @@
+            tzRoundingBuilder = Rounding.builder(tryIntervalAsFixedUnit());
+        } else if (intervalType.equals(DateIntervalWrapper.IntervalTypeEnum.CALENDAR)) {
+            tzRoundingBuilder = Rounding.builder(tryIntervalAsCalendarUnit());
+        } else {
```
The `else` section is basically both `LEGACY` branches, using the old parsing logic.  That's why it retains the "try as calendar first" logic, and then falls back to a fixed interval if it can't parse as a calendar.  

The old logic is setup that way since `LEGACY_DATE_HISTO` can be both calendar and fixed. :(

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java:
```diff
@@ -0,0 +1,415 @@
+        }
+
+        // interval() method
+        switch (type) {
```
> If we're ok breaking DateHistoAggBuilder so that you can only set one of: [interval, dateHistogramInterval, fixedInterval, calendarInterval], it would definitely simplify some of the logic. But I'm not sure we could do that in a minor since it'd break java users (unsure if you can set both in REST)

Ok agreed that we shouldn't break DateHistoAggBuilder here as it would mean a break in a minor. I assume that when we come to removing the legacy interval option this logic will become simple anyway?

polyfractal replied:
Correct, this will be simplified considerably once the legacy options are removed.  Then we can just do the mutual exclusion check on the setters themselves and life should be easy.

I'll add some comments laying out the plans so it's clear

colings86 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/DateHistogramGroupConfig.java:
```diff
@@ -68,9 +93,82 @@
+        }
+    }
+
+    static DateHistogramGroupConfig fromUnknownTimeUnit(String field, DateHistogramInterval interval,
```
Could you add a JavaDocs to this and the method below explaining when this should be used and when the normal constructors should be used? Its not clear to me when it would be appropriate to use the normal constructors?

polyfractal replied:
Sure.  The tl;dr: is this handles the legacy case where jobs are stored with `interval` and we need to "up-convert" that to one of fixed/calendar.  Once the deprecation period is over, we'll remove the default ctor so a user can only choose Fixed or Calendar directly (in java), and remove the option for `interval` in the REST layer.

colings86 reviewed (commented):
I left some comment for the rollup commit.

I also wonder whether we should add some YAML test for the legacy interval which as well as checking that it does the right thing also checks for the deprecation warning? I think these tests should cover fixed and calendar legacy intervals

colings86 commented on rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/250_moving_fn.yml:
```diff
@@ -6,6 +6,10 @@ setup:
 
+  - skip:
+      version: " - 7.99.0"
+      reason:  "calendar_interval added in 7.1"
```
Should we have 2 versions of this test, one for the legacy interval and one for the new intervals?

polyfractal replied:
This test already catches a warning (the bad window), and I don't think the yaml runner allows catching two warnings at the same time.  We'd also need to catch the deprecation warning for the legacy `interval`

It's a good question in general though.  I tried to test both versions in tests about the date_histo, but converted to the "correct" new version for all other tests.  If we want we can add back some of the legacy intervals in some places, or all places.

polyfractal replied:
I misread, the existing catch is for an exception.  I'll add a block for warnings and duplicate the test.

colings86 commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/DateHistogramIT.java:
```diff
@@ -1433,7 +1433,7 @@ public void testDSTEndTransition() throws Exception {
                 .setQuery(new MatchNoneQueryBuilder())
                 .addAggregation(dateHistogram("histo").field("date").timeZone(ZoneId.of("Europe/Oslo"))
-                        .dateHistogramInterval(DateHistogramInterval.HOUR).minDocCount(0).extendedBounds(
+                        .calendarInterval(DateHistogramInterval.HOUR).minDocCount(0).extendedBounds(
```
Maybe we should test the legacy interval here too?

colings86 commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilderTests.java:
```diff
@@ -36,10 +36,11 @@ private DateHistogramValuesSourceBuilder randomDateHistogramSourceBuilder() {
         }
         if (randomBoolean()) {
-            histo.dateHistogramInterval(randomFrom(DateHistogramInterval.days(10),
+            histo.calendarInterval(randomFrom(DateHistogramInterval.days(1),
```
Maybe we should test the legacy interval here too?

polyfractal replied:
I'm not sure how to get the warnings to play nice.  E.g. if this creates a legacy interval, it will throw a deprecation warning in the tests executed by `BaseAggregationTestCase`, so there's no opportunity to `assertWarnings()`.

I could store what kind of interval was chosen in a static variable and then enable/disable warning checks based on that.  A bit better than nothing?

colings86 replied:
Enabling and disabling warnings based on a static variable seems fragile to me. Maybe we should create a new class `LegacyCompositeAggregationBuilderTests` which will build histos with legacy intervals and then we can override the test methods from `BaseAggregationTestCase`, call the super method and then assert the warning. WDYT?

polyfractal replied:
I like that approach.  Seems much saner than my suggestion :)

colings86 commented on docs/reference/aggregations/pipeline/movavg-aggregation.asciidoc.orig:
```diff
@@ -0,0 +1,666 @@
+[[search-aggregations-pipeline-movavg-aggregation]]
```
I don't think you meant to include this file? This looks like leftover from a merge conflict?

colings86 reviewed (commented):
Left a comment ont he docs commit

colings86 reviewed (commented):
@polyfractal I've made it through all the commits. I did review the ML commit but only in so far as to look at the test muting etc. We will obviously want to get those test modified to work with the new intervals. We should also make sure we have some tests on upgrading ML jobs from legacy to new intervals

Commit:
Address review comments for core code commit

polyfractal commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/DateHistogramGroupConfig.java:
```diff
@@ -203,7 +320,7 @@ public boolean equals(final Object other) {
             return true;
         }
-        if (other == null || getClass() != other.getClass()) {
+        if (other == null || other instanceof DateHistogramGroupConfig == false) {
```
This one is a bit tricky.  This is there to accomodate the legacy option.  E.g. `5m` in `DateHistogramGroupConfig` == `5m` in `DateHistogramGroupConfig.FixedInterval`.

I could make the check more explicit and make sure it is either the same class or the super class, which would rule out `Fixed` vs `Calendar` comparisons?

colings86 replied:
> I could make the check more explicit and make sure it is either the same class or the super class, which would rule out Fixed vs Calendar comparisons?

I think we should do this at a minimum as I think its important that when the legacy code is removed we don't have a situation where `Fixed == Calendar` is possible.

I'm wondering though when the comparison between legacy and new is actually needed? Presumably within a node/shard its either going to be interpreted as legacy OR as a new interval but never both? I suppose one place this could break is in the reduce on the coordinating node but do we actually need to do a comparison there?

polyfractal replied:
Thinking about it, I don't think this will ever actually happen in practice.  New nodes will always "upconvert" the old object into a new one (using that "from unknown interval") method.  I don't quite recall but this may have been added because of failing randomized serialization tests or something.  I'll investigate if we actually need it.

If there is a compelling reason for it, I'll add the `Fixed != Calendar` check at the very least.

polyfractal replied:
Yeah dunno, reverted this change and all tests pass.  

This is the rollup datehisto config object, and I don't think it's actually compared anywhere anyway so this leniency isn't needed anywhere from what i can see.  Think it might just have been overthinking the change when I first did it :)  Good catch, thanks!

Commit:
Address review comments for rollup commit

Commit:
Address review comments for testing commit

Commit:
checkstyle

Commit:
Remove unnecessary loosening of equals()

Commit:
Add LegacyIntervalCompositeAggBuilderTests, simplify deprecations

Removes the interval-only deprecation string because it complicates
testing considerably, and is confusing for the user.

Commit:
Merge remote-tracking branch 'origin/master' into date_histo_explicit_intervals

Commit:
Checkstyle

Commit:
Fix rollup translation to use dateHistoInterval when appropriate

Commit:
Add extra test for moving_fn deprecated interval

Commit:
DataFramePivot, composite test fixes

benwtrent reviewed (approved):
The change from `interval(milliseconds)` to `fixed_interval(####ms)` makes sense for our ML stuff. This will also simplify some things we have to handle in ML and DataFrames as we will not have to support both `long` and `calendarInterval` in the same fields. 

A huge help would be flagging things as TODO, or creating a new issue flagged as `:ml` so that we remember to pick this up and fix up the rest of the missing pieces.

colings86 reviewed (approved):
The changes LGTM. Happy for this to be merged if there are no significant changes needed to get the build passing

Commit:
Add deprecation warnings to getters, re-arrange checks to avoid warning

Commit:
Merge remote-tracking branch 'origin/master' into date_histo_explicit_intervals

Commit:
Test fixes

Commit:
checkstyle! argh

polyfractal commented:
And we're green!
 
![adcc08a87d810d4d](https://user-images.githubusercontent.com/1224228/57178573-977c9f80-6e40-11e9-8876-e111d15eb97f.gif)

I'll merge this on monday and put up a backport PR shortly after (IIRC there are some tweaks needed for the backport).

Commit:
Bump deprecation version

polyfractal commented:
@elasticmachine run elasticsearch-ci/1 
@elasticmachine run elasticsearch-ci/default-distro

polyfractal commented:
@elasticmachine run elasticsearch-ci/1

polyfractal commented:
https://github.com/elastic/elasticsearch/pull/41906 has merged, removing backport label and this is done! :tada:
